multiview/embed/splitae.py

```diff
@@ -0,0 +1,206 @@
+import torch
+from torch.utils.data import Dataset, DataLoader
+from torchvision import datasets
+import torchvision
+import matplotlib.pyplot as plt
+import PIL
+import numpy as np
+import itertools
+import tqdm
+
+from multiview.embed.base import BaseEmbed
+
+
+class FullyConnectedNet(torch.nn.Module):
+    def __init__(self, inputSize, hiddenSize, numHiddenLayers, embeddingSize):
+        super().__init__()
+        assert numHiddenLayers >= 0, "can't have negative hidden layer count"
+        self.layers = torch.nn.ModuleList()
+        if numHiddenLayers == 0:
+            self.layers.append(torch.nn.Linear(inputSize, embeddingSize))
+        else:
+            self.layers.append(torch.nn.Linear(inputSize, hiddenSize))
+            for i in range(numHiddenLayers-1):
+                self.layers.append(torch.nn.Linear(hiddenSize, hiddenSize))
+            self.layers.append(torch.nn.Linear(hiddenSize, embeddingSize))
+
+    def forward(self, x):
+        for layer in self.layers[:-1]:
+            x = torch.nn.Sigmoid()(layer(x))
+        x = self.layers[-1](x)  # no activation on last layer
+        return x
+
+    def paramCount(self):
+        return np.sum([np.prod(s.shape) for s in self.parameters()])
+
+
+class SplitAE(BaseEmbed):
+    """
+    Implements an autoencoder that creates an embedding of a view View1 and
+    from that embedding reconstructs View1 and another view View2.
+    Parameters
+    ----------
+    hiddenSize: number of nodes in the hidden layers
+    numHiddenLayers: number of hidden layers in each encoder or
+        decoder net
+    embedSize: size of the bottleneck vector in the autoencoder
+    trainingEpochs: how many times the network trains on the full
+        dataset
+    learningRate: learning rate of the Adam optimizer
+    Attributes
+    ----------
+    view1Encoder: the View1 embedding network as a PyTorch module
+    view1Decoder: the View1 decoding network as a PyTorch module
+    view2Decoder: the View2 decoding network as a PyTorch module
+    """
+
+    def __init__(self, hiddenSize=64, numHiddenLayers=2, embedSize=20,
+                 trainingEpochs=10, batchSize=16, learningRate=0.001):
+        self.hiddenSize = hiddenSize
+        self.embedSize = embedSize
+        self.numHiddenLayers = numHiddenLayers
+        self.trainingEpochs = trainingEpochs
+        self.batchSize = batchSize
+        self.learningRate = learningRate
+
+    # Xs is not a tensor but instead a list with two arrays of shape [n, f_i]
+    def fit(self, Xs, validationXs=None, printInfo=True):
+        """
+        Given two views, create and train the autoencoder.
+        Parameters
+        ----------
+        Xs: a list with two arrays. Each array has `n` rows (samples) and some
```

Comment:
Maybe make this consistent wording with how @rflperry started doing it, just so our package is as uniform as possible. https://github.com/NeuroDataDesign/multiview/pull/45/files